Code/GraphMol/FindRings.cpp

```diff
@@ -715,69 +711,52 @@ int BFSWorkspace::smallestRingsBfs(const ROMol &mol, int root,
-              com++;
-              id = (*ci);
-              if (id != root) {
-                break;
-              }
-            }
-          }  // end of found stuff in common with neighbor
-
-          if (id == root) {  // we found a ring
-            // make the ring
-            INT_VECT ring = cpath;
-
-            // remove the root node and attach the other half of the ring from
-            // npath
-            // reverse this piece so that the ring is traversed correctly
-
-            // FIX: we're probably assured that root is the first node, so we
-            // can
-            //  just pop it from the front
-            npath.erase(std::remove(npath.begin(), npath.end(), root));
-
-#ifndef WIN32
-            ring.insert(ring.end(), npath.rbegin(), npath.rend());
-#else  // I <heart> MSVC++ v6
-            std::reverse(npath.begin(), npath.end());
-            ring.insert(ring.end(), npath.begin(), npath.end());
-#endif
-            if (ring.size() <= curSize) {
-              curSize = rdcast<unsigned int>(ring.size());
-              rings.push_back(ring);
-            } else {
-              // we are done with the smallest rings
-              return rdcast<unsigned int>(rings.size());
-            }
-          }  // end of found a ring
-        }    // end of we have seen this neighbor before
-      }      // end of nbrIdx not part of current path and not a done atom
+      if (d_done[nbrIdx] == BLACK || d_parents[curr] == nbrIdx) {
+        continue;
+      }
+      if (d_done[nbrIdx] == WHITE) {
+        // we have never been to this node before through via any path
+        d_parents[nbrIdx] = curr;
+        d_done[nbrIdx] = GRAY;
+        d_depths[nbrIdx] = depth;
+        bfsq.push_back(nbrIdx);
+      } else {
+        // we have been here via a different path
+        // there is a potential for ring closure here
+        // stitch together the two paths
+
+        ring = {nbrIdx};
+        // forwards path
+        int parent = d_parents[nbrIdx];
+        while (parent != -1 && parent != root) {
+          ring.push_back(parent);
+          parent = d_parents[parent];
+        }
+
+        // backwards path
+        ring.insert(ring.begin(), curr);
+        parent = d_parents[curr];
+        while (parent != -1) {
+          // Is the least common ancestor not the root?
+          if (std::find(ring.begin(), ring.end(), parent) != ring.end()) {
+            ring.clear();
+            break;
+          }
+          ring.insert(ring.begin(), parent);
+          parent = d_parents[parent];
+        }
+
+        // Found a new small ring including the root.
+        if (ring.size() > 1) {
+          if (ring.size() <= curSize) {
+            curSize = rdcast<unsigned int>(ring.size());
+            rings.push_back(ring);
+          } else {
+            // we are done with the smallest rings
+            return rdcast<unsigned int>(rings.size());
+          }
+        }
```

Comment:
The order of the indices in the ring are going to be based on the order of the atoms in the bond arrays, right? Like, this construction will guarantee that the atom indices are in order _if_ the bond arrays are in order.

I wonder if it would make sense to add a call to order the atom indices to ensure that we always get the same answer going forward.